Error building with glibc 2.33 because of _STAT_VER changes

The glibc maintainers have done something with the `stat` wrappers in 2.33: https://github.com/bminor/glibc/commit/8ed005daf0ab03e142500324a34087ce179ae78e

While running `make` to build `ekam` under glibc 2.33:
```
✘ learn: ekam/rules/intercept.ekam-rule
    src/ekam/rules/intercept.c: In function ‘direct_stat’:
    src/ekam/rules/intercept.c:671:21: error: ‘_STAT_VER’ undeclared (first use in this function)
      671 |   return real_xstat(_STAT_VER, path, sb);
          |                     ^~~~~~~~~
    src/ekam/rules/intercept.c:671:21: note: each undeclared identifier is reported only once for each function it appears in
    src/ekam/rules/intercept.c:672:1: warning: control reaches end of non-void function [-Wreturn-type]
      672 | }
          | ^
```

The `intercept.c` line in question: https://github.com/capnproto/ekam/blob/454152e658db071f62a45eae620ef244c1726165/src/ekam/rules/intercept.c#L671

Note that `ekam` previously built on the same Linux install but a glibc upgrade broke compilation as seen above, so I suspect this issue is coming to Debian/Ubuntu eventually.

For now, I built an `ekam` binary in a Docker container using an older glibc, and I replaced the `_STAT_VER` in my host system's `intercept.c` with `3` and things seem to work. I figured I'd put this up now for anyone else Googling. I'm a bit out of my depth on what's going on in that glibc change, or why `_STAT_VER` didn't make it into _anything_ under my `/usr/include` (I have no `xstatver.h`, for example), but I'll try to poke at it some more in a bit.

--- a/src/ekam/rules/intercept.c
+++ b/src/ekam/rules/intercept.c
@@ -594,6 +594,43 @@ int ___lxstat64 (int ver, const char* path, struct stat64* sb) {
   return __lxstat64 (ver, path, sb);
 }
 
+#ifndef _STAT_VER
+/* glibc 2.33+ no longer defines `stat()` as an inline wrapper in the header, but instead exports
+ * it as a regular symbol. So, we need to intercept it. And, we need to give `lstat()` special
+ * treatment just like `__lxstat()` above. */
+
+WRAP(int, stat, (const char* path, struct stat* sb), (path, sb), READ, -1)
+//WRAP(int, lstat, (const char* path, struct stat* sb), (path, sb), READ, -1)
+
+typedef int lstat_t (const char* path, struct stat* sb);
+int lstat(const char* path, struct stat* sb) {
+  static lstat_t* real_lstat = NULL;
+  static stat_t* real_stat = NULL;
+  char buffer[PATH_MAX];
+
+  if (real_lstat == NULL) {
+    real_lstat = (lstat_t*) dlsym(RTLD_NEXT, "lstat");
+    assert(real_lstat != NULL);
+  }
+  if (real_stat == NULL) {
+    real_stat = (lstat_t*) dlsym(RTLD_NEXT, "stat");
+    assert(real_stat != NULL);
+  }
+
+  path = remap_file("lstat", path, buffer, READ);
+  if (path == NULL) return -1;
+  if (path[0] == '/') {
+    return real_lstat(path, sb);
+  } else {
+    return real_stat(path, sb);
+  }
+}
+int _lstat(const char* path, struct stat* sb) {
+  return lstat(path, sb);
+}
+
+#endif  /* defined(_STAT_VER) */
+
 /* Cannot intercept intra-libc function calls on Linux, so we must intercept fopen() as well. */
 WRAP(FILE*, fopen, (const char* path, const char* mode), (path, mode),
      mode[0] == 'w' || mode[0] == 'a' ? WRITE : READ, NULL)
@@ -661,6 +698,8 @@ int _execvp (const char* path, char* const argv[]) {
 
 /* Called by access(), below. */
 static int direct_stat(const char* path, struct stat* sb) {
+#ifdef _STAT_VER
+  // glibc <2.33 defines `stat()` as an inline wrapper in the header file, which calls __xstat().
   static __xstat_t* real_xstat = NULL;
 
   if (real_xstat == NULL) {
@@ -669,6 +708,18 @@ static int direct_stat(const char* path, struct stat* sb) {
   }
 
   return real_xstat(_STAT_VER, path, sb);
+#else  /* defined(_STAT_VER) */
+  // glibc 2.33+ no longer defines stat() as an inline wrapper in the header, but instead exports
+  // it as a regular symbol.
+  static stat_t* real_stat = NULL;
+
+  if (real_stat == NULL) {
+    real_stat = (stat_t*) dlsym(RTLD_NEXT, "stat");
+    assert(real_stat != NULL);
+  }
+
+  return real_stat(path, sb);
+#endif  /* defined(_STAT_VER), #else */
 }
 
 typedef ssize_t readlink_t(const char *path, char *buf, size_t bufsiz);

